wasm/wasm_source/src/vp_user.rs

```diff
@@ -122,12 +126,85 @@ fn validate_tx(
+                    Some(address) => *address == addr && *valid_sig,
+                    None => true,
+                };
+
+                // Changes due to unjailing, deactivating, and reactivating are
+                // marked by changes in validator state
+                let state_change =
+                    proof_of_stake::storage::is_validator_state_key(key);
+                let valid_state_change = match state_change {
+                    Some((address, epoch)) => {
+                        let params_pre =
+                            proof_of_stake::read_pos_params(&ctx.pre())?;
+                        let state_pre =
+                            proof_of_stake::validator_state_handle(address)
+                                .get(&ctx.pre(), epoch, &params_pre)?;
+
+                        let params_post =
+                            proof_of_stake::read_pos_params(&ctx.post())?;
+                        let state_post =
+                            proof_of_stake::validator_state_handle(address)
+                                .get(&ctx.post(), epoch, &params_post)?;
+
+                        match (state_pre, state_post) {
+                            (Some(pre), Some(post)) => {
+                                if
+                                // Deactivation case
+                                (matches!(
+                                    pre,
+                                    ValidatorState::Consensus
+                                        | ValidatorState::BelowCapacity
+                                        | ValidatorState::BelowThreshold
+                                ) && post == ValidatorState::Inactive)
+                                // Reactivation case
+                                || pre == ValidatorState::Inactive
+                                    && post != ValidatorState::Inactive
+                                // Unjail case
+                                || pre == ValidatorState::Jailed
+                                    && matches!(
+                                        post,
+                                        ValidatorState::Consensus
+                                            | ValidatorState::BelowCapacity
+                                            | ValidatorState::BelowThreshold
+                                    )
+                                {
+                                    *address == addr && *valid_sig
+                                } else {
+                                    true
+                                }
+                            }
+                            (None, Some(_post)) => {
+                                // Becoming a validator must be authorized
+                                *valid_sig
+                            }
+                            _ => true,
+                        }
+                    }
+                    None => true,
+                };
+
+                valid_bond_or_unbond_change
```

Comment:
if the change is unrecognized, shouldn't we reject it by default? looks like we accept it by default here

Reply:
ah, yes, we decided to do this in https://github.com/anoma/namada/issues/408. Let's deal with this separately though as it affects other sub-systems too

Reply:
> I think not checking unbonds is fine in principle, we should have a quick test for this though to make sure nothing breaks.

I think if anything breaks it will most likely be interactions with some other thing like slashing or rewards (though to my understanding it is safe - see https://github.com/anoma/namada/pull/2208#discussion_r1403114572). I'd also like to test this and imho the best way to check this would be to add this to state machine, but we don't yet cover rewards and couple other things